### ISSUE TITLE
Big strings fail on Cairo (linux only?)

### ISSUE THREAD

**Issue by erg** [factor/factor#547]
``` factor
10000 CHAR: a <string> print

An error occurred while drawing the world T{ world f ~array~ ~array~ f f ~vector~ ~array~....
This world has been deactivated to prevent cascading errors.
cairo-error
message
"invalid value (typically too big) for the size of the input (..."
```

**Comment by kusumotonorio** [comment]
I have just started the experiment and am in the process of confirming the concept, but I would appreciate if you could give me your opinion.

If we adopt the following solution, are we likely to encounter problems?
- Split long strings into non-problematic lengths with graphene breaks.
- `print` (`write`) split strings sequentially

The following code is for `print`, but I think `stream-write` should be changed. We might also want to limit our changes to methods that image strings.

```Factor
USING: locals unicode strings ;
:: stream-print' ( str stream -- )
    str :> s!
    s length :> len!
    0 :> end!
    [
        len 4681 > [ 4681 end! ] [ len end! ] if
        end s last-grapheme-from end!
        s end 1 + cut-slice :> ( before-s after-s )
        before-s >string stream stream-write
        after-s length 0 = [ f ] [
            after-s s!
            s length len!
            t
        ] if
    ] loop
    stream stream-nl ;

: print' ( str -- ) output-stream get stream-print' ; inline
```

![linuxFactor](https://user-images.githubusercontent.com/47816570/73130994-65524400-4046-11ea-9434-2f1b9ed63262.png)

**Comment by erg** [comment]
I have never really looked into it, but really long strings are extremely
slow. Every platform.

**Comment by kusumotonorio** [comment]
It's surprising, but it seems to take less time to print in pieces.

![print print'](https://user-images.githubusercontent.com/47816570/73134263-e118b480-4077-11ea-8682-791f9ef6ba2c.png)

**Comment by mrjbq7** [comment]
We should make that change as low level as we can, instead of assuming all streams have that restriction. 

Maybe in pane-stream where it’s converted to an image, or perhaps we can make strings backed by multiple images and hide the complexity in the backend.

**Comment by kusumotonorio** [comment]
@mrjbq7 I moved the experimental code to `M: pane-stream stream-write`.
The code seems to work well on a limited number of tests, but it needs to be considered more. For example, my current code cannot avoid breaking escape sequences.

I agree that pushing this kind of processing to the backend is the right thing to do, but given how long this problem has not been solved, I think it's a good idea to do it in a simple way now.

**Comment by mrjbq7** [comment]
That makes sense.

Are you thinking of setting a number of code points to break on based on a
minimal number that works across all platforms?

**Comment by kusumotonorio** [comment]
Yes. It's, as far as I can tell, 3639 for 32 bit Windows. 
We can set a constant for each platform, but it doesn't seem to be very common to display so many characters in a gadget, and it doesn't seem to be slow to split up, so I think that's fine.

**Comment by kusumotonorio** [comment]
> The code seems to work well on a limited number of tests, but it needs to be considered more. For example, my current code cannot avoid breaking escape sequences.

I confused the literal representation of strings with the sequence of codepoints. The changed code should not have to deal with escape codes especially.

I wrote the following code:

https://github.com/factor/factor/compare/master...kusumotonorio:stream-write

**Comment by mrjbq7** [comment]
This might be equivalent, but splits into pieces, then writes each piece.

```factor
M: pane-stream stream-write
    [
        [
            [ length dup 3639 > ] keep
            '[ drop 3639 _ last-grapheme-from ] when
        ] >pieces [ "" like split-lines ] map
    ] dip [ '[ _ pane-write ] each ] do-pane-stream ;
```

**Comment by kusumotonorio** [comment]
Your code is more elegant!

Is there a difference in memory usage?
If the GC doesn't work until the end of the execution of these words, then my code has no advantage, since it has to make all of the real partitioned data after all.

**Comment by mrjbq7** [comment]
If we want to minimize memory usage, then this keeps them all as slices, then makes each one concrete strings-to-lines and writes, allowing each to be eligible for garbage collection immediately.

```factor
M: pane-stream stream-write
    [
        [
            [ length dup 3639 > ] keep
            '[ drop 3639 _ last-grapheme-from ] when
        ] >pieces
    ] dip [
        '[ "" like split-lines _ pane-write ] each
    ] do-pane-stream ;
```

**Comment by mrjbq7** [comment]
One disadvantage of my code above is that in the common case, it will copy the string, so maybe we should do something like this, which is more similar to your code:

```factor
M: pane-stream stream-write
    [
        '[
            dup length 3639 >
            [ 3639 over last-grapheme-from cut-slice ] [ f ] if
            swap "" like split-lines _ pane-write dup
        ] loop drop
    ] do-pane-stream ;
```
```

**Comment by kusumotonorio** [comment]
`"http://www.google.com" http-get nip write` and `100,000 CHAR: a <string> write` work fine on 32bit Wndows7.

If you see how your code works, please do PR and merge it.

**Comment by mrjbq7** [comment]
Fixed in 94cbe2b4796fd97d99e8ab32244f1fd3f7362721.

Also fixes ``stream-format`` which had the same problem.

**Comment by mrjbq7** [comment]
Thanks @kusumotonorio !
